### PR TITLE
Microservices and configuration

### PR DESCRIPTION
**Pull request by kevinhodges**
### Blog title
Micro-services and configuration, a different approach.

### Blog description
How we solved our single source of truth issue in a microservices environment.

#### Author
- [x] I have checked for grammatical errors
- [x] I have checked for spelling errors
- [x] All information (technical or otherwise) is correct to my knowledge

#### Editor 1
- [x] I have checked for grammatical errors
- [x] I have checked for spelling errors
- [x] All information (technical or otherwise) is correct to my knowledge

#### Editor 2
- [x] I have checked for grammatical errors
- [x] I have checked for spelling errors
- [x] All information (technical or otherwise) is correct to my knowledge

### CODE REVIEW

File: _posts/2015-12-18-microservices-and-configuration.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+---
+layout: post
+title: Micro-services and configuration, a different approach.
+categories: NodeJS
+author: kevin_hodges
+tags: microservices, document store, couchdb, pouchdb, configuration management
+comments: true
+---
+
+Moving to a micro-services architecture is the dream right?  We’ve been breaking up our monolith at HX for a few years now and winning in all sorts of areas from better code reuse to greater decoupling of functionality to quicker release cycles.  We did very quickly however come a cross an elephant in the room that no one wanted to talk about.  It was a by-product or our micro-services journey and things got complicated very quickly.
```
Comment (notABluesSinger) [inline _posts/2015-12-18-microservices-and-configuration.md]: Typo: by-product **or** our

File: _posts/2015-12-18-microservices-and-configuration.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+On the monolith it was pretty straightforward.  It’s all in one place and "on box".  Now we’ve got many boxes all doing different things that are in different parts of the world, keeping configuration consistent and available across them all is a bit of a headache, especially as you want to keep a [single source of truth](https://en.wikipedia.org/wiki/Single_Source_of_Truth).
+
+There were a few architectural points we had to guarantee when looking for a config solution.
+
+1. Services can't go “off box” for it.  Round trips to other services can be expensive (http/latency), especially for high volume calls such as configuration.
+2. We have to have a single source of truth.
+3. It needs to be editable and changes made available in real time (no deploying of services to see the updates).
+
+The solution we came up with simple and quite elegant.  Hosted CouchDB as the central store and [PouchDB](http://pouchdb.com/) running on every service.
```
Comment (notABluesSinger) [inline _posts/2015-12-18-microservices-and-configuration.md]: I think you're missing a word here;
'The solution we came up with simple' > 'The solution we came up with is simple'

**Comment by notABluesSinger** [comment]
There's a few double spaces throughout, pedantic I know... sorry.

**Comment by kevinhodges** [comment]
@notABluesSinger fixes made

**Comment by notABluesSinger** [comment]
:+1:

**Comment by hx-markterry** [comment]
taken

File: _posts/2015-12-18-microservices-and-configuration.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+---
+layout: post
+title: Micro-services and configuration, a different approach.
+categories: NodeJS
+author: kevin_hodges
+tags: microservices, document store, couchdb, pouchdb, configuration management
+comments: true
+---
+
+Moving to a micro-services architecture is the dream right? We’ve been breaking up our monolith at HX for a few years now and winning in all sorts of areas from better code reuse to greater decoupling of functionality to quicker release cycles. We did very quickly however come a cross an elephant in the room that no one wanted to talk about. It was a by-product of our micro-services journey and things got complicated very quickly.
```
Comment (hx-markterry) [inline _posts/2015-12-18-microservices-and-configuration.md]: would change:

`from better code reuse to greater decoupling of functionality to quicker release cycles`

to

`such as better code reuse, greater decoupling of functionality and quicker release cycles`

File: _posts/2015-12-18-microservices-and-configuration.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+Config.
```
Comment (hx-markterry) [inline _posts/2015-12-18-microservices-and-configuration.md]: Maybe make this a heading or at least bold etc.?

File: _posts/2015-12-18-microservices-and-configuration.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+2. We have to have a single source of truth.
+3. It needs to be editable and changes made available in real time (no deploying of services to see the updates).
+
+The solution we came up with is simple and quite elegant. Hosted CouchDB as the central store and [PouchDB](http://pouchdb.com/) running on every service.
+
+We took [Cloudant](https://cloudant.com) off the shelf and gave them some money, we don’t want to be writing anything that other people can do faster and better. All our config is json and no where near “relational” so a document store suits this model well. It also has a tidy interface to allow people to make changes to config easily. We then use PouchDB as an npm module on the server. Each service when it boots up fires up our configuration module which is a simple Pouch wrapper. Pouch takes care of the sync'ing which ensures that if changes are made in Cloudant, they are immediately pulled down to maintain a complete config set. A byproduct of this is, if for some reason Cloudant becomes unavailable, a copy of the configuration is in memory to ensure services can continue to operate.
```
Comment (hx-markterry) [inline _posts/2015-12-18-microservices-and-configuration.md]: Do we need to mention why cloudant?

Comment (hx-markterry) [inline _posts/2015-12-18-microservices-and-configuration.md]: `json` should be written as `JSON` and below too.

File: _posts/2015-12-18-microservices-and-configuration.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+We’ve been running in production with this for over a year now and starting to scale up and have had zero problems with availability. The main task is keeping json configs organised well. We don’t have the luxury of nested directories to group configuration as our document store provides a simple flat structure so ensuring that remains tidy and visible is something we’re working hard on with documentation and peer reviews.
```
Comment (hx-markterry) [inline _posts/2015-12-18-microservices-and-configuration.md]: I woud change:

`The main task is keeping json configs organised well.`

to:

`The main task is keeping JSON configs well organised.`

**Comment by hx-markterry** [comment]
is it worth giving an example of the flat config structure in couchdb?

**Comment by hx-markterry** [comment]
Looking good.

**Comment by hx-markterry** [comment]
send it! :rocket:

**Comment by hpoom** [comment]
One change `All our config is JSON and no where near “relational” ` should read `nowhere`, not `no where`

**Comment by hpoom** [comment]
Damn it, did not mean to close :cry: Stup Github buttons

**Comment by hpoom** [comment]
Great blog post, :shipit:

**Pull request merged**